dhall/src/Dhall/Import.hs

```diff
@@ -510,20 +526,25 @@ loadImportWithSemanticCache
             let actualHash = Dhall.Crypto.sha256Hash bytesStrict
 
             if semanticHash == actualHash
-                then return ()
-                else do
-                    Status { _stack } <- State.get
-                    throwMissingImport (Imported _stack (HashMismatch {expectedHash = semanticHash, ..}))
+                then do
+                    let bytesLazy = Data.ByteString.Lazy.fromStrict bytesStrict
 
-            let bytesLazy = Data.ByteString.Lazy.fromStrict bytesStrict
+                    importSemantics <- case Dhall.Binary.decodeExpression bytesLazy of
+                        Left  err -> throwMissingImport (Imported _stack err)
+                        Right e   -> return e
 
-            importSemantics <- case Dhall.Binary.decodeExpression bytesLazy of
-                Left  err -> throwMissingImport (Imported _stack err)
-                Right e   -> return e
+                    return (ImportSemantics {..})
+                else do
+                    printWarning $
+                        makeHashMismatchMessage semanticHash actualHash
+                        <> "\n"
+                        <> "The interpreter has automatically fixed cached import\n"
+                    saveCacheToFile
 
-            return (ImportSemantics {..})
 
-        Nothing -> do
+        Nothing -> saveCacheToFile
+    where
+        saveCacheToFile = do
```

Comment:
There's a lot more going on here than just writing a new cache file. Not sure how to best capture all that in a name. Maybe just use something generic like `fetch`?